boot/seal.go

```diff
@@ -517,57 +525,78 @@ func resealFallbackObjectKeys(pbc predictableBootChains, authKeyFile string, rol
 
 // TODO:UC20: this needs to take more than one model to accommodate the remodel
 // scenario
-func recoveryBootChainsForSystems(systems []string, trbl bootloader.TrustedAssetsBootloader, modeenv *Modeenv) (chains []bootChain, err error) {
-	model := modeenv.ModelForSealing()
-	for _, system := range systems {
-		// get kernel and gadget information from seed
-		perf := timings.New(nil)
-		_, snaps, err := seedReadSystemEssential(dirs.SnapSeedDir, system, []snap.Type{snap.TypeKernel, snap.TypeGadget}, perf)
-		if err != nil {
-			return nil, fmt.Errorf("cannot read system %q seed: %v", system, err)
-		}
-		if len(snaps) != 2 {
-			return nil, fmt.Errorf("cannot obtain recovery system snaps")
-		}
-		seedKernel, seedGadget := snaps[0], snaps[1]
-		if snaps[0].EssentialType == snap.TypeGadget {
-			seedKernel, seedGadget = seedGadget, seedKernel
-		}
+func recoveryBootChainsForSystems(systems []string, trbl bootloader.TrustedAssetsBootloader, modeenv *Modeenv, includeTryModel bool) (chains []bootChain, err error) {
+
+	chainsForModel := func(model secboot.ModelForSealing) error {
+		modelID := modelUniqueID(model)
+		for _, system := range systems {
+			// get kernel and gadget information from seed
+			perf := timings.New(nil)
+			seedSystemModel, snaps, err := seedReadSystemEssential(dirs.SnapSeedDir, system, []snap.Type{snap.TypeKernel, snap.TypeGadget}, perf)
+			if err != nil {
+				return fmt.Errorf("cannot read system %q seed: %v", system, err)
+			}
+			if len(snaps) != 2 {
+				return fmt.Errorf("cannot obtain recovery system snaps")
+			}
+			seedModelID := modelUniqueID(seedSystemModel)
```

Comment:
thanks for the change, could you add a TODO here that is not of immediate concern though:

// TODO: consider relaxing this to ignore the ModelSignKeyID when matching models,
// OTOH we would need to properly take into account key expiration and revocation